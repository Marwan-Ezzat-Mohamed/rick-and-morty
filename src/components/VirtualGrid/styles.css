.virtual-grid > div {
  position: relative !important;
}
