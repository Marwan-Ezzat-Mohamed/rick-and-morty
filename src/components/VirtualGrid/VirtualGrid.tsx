import React, { useRef } from 'react';
import AutoSizer from 'react-virtualized-auto-sizer';
import InfiniteLoader from 'react-window-infinite-loader';
import LinearProgress from '@mui/material/LinearProgress';
import {
  FixedSizeGrid as Grid,
  GridChildComponentProps,
  areEqual,
} from 'react-window';
import './styles.css';

type ItemCardGridProps<Item> = {
  items: Item[] | null;
  ItemCard: React.FC<Item>;
  itemHeight: number;
  itemWidth: number;
  gap?: number;
  loadMoreItems: (
    startIndex: number,
    stopIndex: number,
  ) => Promise<void> | void;
  isItemLoaded: (index: number) => boolean;
};

const VirtualGrid = <Item,>({
  items,
  ItemCard,
  itemHeight,
  itemWidth,
  gap = 0,
  loadMoreItems,
  isItemLoaded,
}: ItemCardGridProps<Item>): JSX.Element => {
  const numberOfItemsThatFits = useRef(0);
  const getIdByGridPosition = (col: number, row: number) =>
    row * numberOfItemsThatFits.current + col;

  if (!items) return <LinearProgress color="inherit" />;
  if (items.length === 0)
    return (
      <h1
        style={{
          textAlign: 'center',
        }}
      >
        No results found
      </h1>
    );
  const Cell = React.memo(
    ({
      columnIndex,
      rowIndex,
      style,
    }: GridChildComponentProps): JSX.Element => {
      const index = getIdByGridPosition(columnIndex, rowIndex);
      return (
        <div style={style}>
          <ItemCard {...items[index]} style={{}} />
        </div>
      );
    },
    areEqual,
  );

  return (
    <AutoSizer>
      {({ height, width }) => {
        numberOfItemsThatFits.current = Math.max(
          Math.floor(width / (itemWidth + gap)),
          1,
        );
        return (
          <InfiniteLoader
            isItemLoaded={isItemLoaded}
            itemCount={(items?.length ?? 0) + 10}
            loadMoreItems={loadMoreItems}
            threshold={5}
          >
            {({ onItemsRendered, ref }) => {
              return (
                <Grid
                  columnCount={Math.min(
                    numberOfItemsThatFits.current,
                    (items?.length ?? 0) + 10,
                  )}
                  columnWidth={itemWidth + gap}
                  height={height}
                  rowCount={
                    ((items?.length ?? 0) + 10) / numberOfItemsThatFits.current
                  }
                  rowHeight={itemHeight + gap}
                  ref={ref}
                  width={width}
                  style={{
                    display: 'flex',
                    justifyContent: 'center',
                    position: 'absolute',
                    paddingTop: gap + 'px',
                  }}
                  className="virtual-grid"
                  onItemsRendered={gridProps => {
                    onItemsRendered({
                      overscanStartIndex:
                        gridProps.overscanRowStartIndex *
                        numberOfItemsThatFits.current,
                      overscanStopIndex:
                        gridProps.overscanRowStopIndex *
                        numberOfItemsThatFits.current,
                      visibleStartIndex:
                        gridProps.visibleRowStartIndex *
                        numberOfItemsThatFits.current,
                      visibleStopIndex:
                        gridProps.visibleRowStopIndex *
                        numberOfItemsThatFits.current,
                    });
                  }}
                >
                  {Cell}
                </Grid>
              );
            }}
          </InfiniteLoader>
        );
      }}
    </AutoSizer>
  );
};

export default VirtualGrid;
