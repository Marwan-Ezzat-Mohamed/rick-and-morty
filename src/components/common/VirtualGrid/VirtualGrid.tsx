import React, { useRef, useCallback } from 'react';
import AutoSizer from 'react-virtualized-auto-sizer';
import InfiniteLoader from 'react-window-infinite-loader';
import LinearProgress from '@mui/material/LinearProgress';
import {
  FixedSizeGrid as Grid,
  GridChildComponentProps,
  areEqual,
} from 'react-window';
import Typography from '@mui/material/Typography';
import './styles.css';

type ItemCardGridProps<Item> = {
  items: Item[] | null;
  ItemCard: React.FC<Item>;
  itemHeight: number;
  itemWidth: number;
  gap?: number;
  loadMoreItems: (
    startIndex: number,
    stopIndex: number,
  ) => Promise<void> | void;
  isItemLoaded: (index: number) => boolean;
  hasMore: boolean;
  numberOfSkeletonsRows?: number;
};

const VirtualGrid = <Item,>({
  items,
  ItemCard,
  itemHeight,
  itemWidth,
  loadMoreItems,
  isItemLoaded,
  hasMore,
  gap = 0,
  numberOfSkeletonsRows = 1,
}: ItemCardGridProps<Item>): JSX.Element => {
  const numberOfColumns = useRef(0);
  const getIdByGridPosition = useCallback(
    (col: number, row: number) => row * numberOfColumns.current + col,
    [],
  );

  if (!items) return <LinearProgress color="inherit" />;

  if (items.length === 0) {
    return (
      <Typography align="center" variant="h1">
        No results found
      </Typography>
    );
  }

  const Cell = React.memo(
    ({
      columnIndex,
      rowIndex,
      style,
    }: GridChildComponentProps): JSX.Element | null => {
      const index = getIdByGridPosition(columnIndex, rowIndex);
      if (!hasMore && index >= items.length) return null;
      return (
        <div style={style}>
          <ItemCard {...items[index]} style={{}} />
        </div>
      );
    },
    areEqual,
  );

  return (
    <AutoSizer>
      {({ height, width }) => {
        const totalNumberOfCards = (items?.length ?? 0) + (hasMore ? 1 : 0);
        numberOfColumns.current = Math.max(
          Math.floor(width / (itemWidth + gap)),
          1,
        );
        const numberOfRows =
          Math.ceil(totalNumberOfCards / numberOfColumns.current) +
          (hasMore ? numberOfSkeletonsRows : 0);

        return (
          <InfiniteLoader
            isItemLoaded={isItemLoaded}
            itemCount={totalNumberOfCards}
            loadMoreItems={loadMoreItems}
          >
            {({ onItemsRendered, ref }) => {
              return (
                <Grid
                  columnCount={Math.min(
                    numberOfColumns.current,
                    totalNumberOfCards,
                  )}
                  columnWidth={itemWidth + gap}
                  height={height}
                  rowCount={numberOfRows}
                  rowHeight={itemHeight + gap}
                  ref={ref}
                  width={width}
                  style={{
                    display: 'flex',
                    justifyContent: 'center',
                    position: 'absolute',
                    marginTop: gap + 'px',
                  }}
                  className="virtual-grid"
                  onItemsRendered={({
                    visibleRowStartIndex,
                    visibleColumnStartIndex,
                    visibleRowStopIndex,
                    overscanRowStopIndex,
                    overscanRowStartIndex,
                  }) => {
                    onItemsRendered({
                      overscanStartIndex:
                        overscanRowStartIndex * numberOfColumns.current,
                      overscanStopIndex:
                        overscanRowStopIndex * numberOfColumns.current,
                      visibleStartIndex:
                        visibleRowStartIndex * numberOfColumns.current,
                      visibleStopIndex:
                        visibleRowStopIndex * numberOfColumns.current,
                    });
                  }}
                >
                  {Cell}
                </Grid>
              );
            }}
          </InfiniteLoader>
        );
      }}
    </AutoSizer>
  );
};

export default VirtualGrid;
