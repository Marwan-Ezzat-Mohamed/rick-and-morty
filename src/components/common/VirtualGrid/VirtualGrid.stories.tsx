import { ComponentStory, ComponentMeta } from '@storybook/react';

import VirtualGrid from '.';

import { Box } from '@mui/material';

export default {
  title: 'Reusable Components/Virtual Grid',
  component: VirtualGrid,
} as ComponentMeta<typeof VirtualGrid>;

const gridProps = {
  itemHeight: 100,
  itemWidth: 100,
  gap: 1,
  loadMoreItems: () => {
    console.log('load more items');
  },
  isItemLoaded: (index: number) => true,
  hasMore: true,
};

const Card = ({ index }: { index: number }) => {
  return (
    <Box
      sx={{
        height: 50,
        width: 25,
        backgroundColor: 'red',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        flex: 1,
      }}
    >
      {index}
    </Box>
  );
};

const mockItems = Array.from({ length: 1000 }).map((_, index) => ({
  index: index,
}));

const PrimaryTemplate: ComponentStory<typeof VirtualGrid> = args => (
  <Box
    sx={{
      height: '100vh',
      display: 'flex',
      flexDirection: 'column',
    }}
  >
    <VirtualGrid {...gridProps} items={mockItems} ItemCard={Card} />
  </Box>
);

export const Primary = PrimaryTemplate.bind({});

const LoadingTemplate: ComponentStory<typeof VirtualGrid> = args => (
  <Box
    sx={{
      height: '100vh',
      display: 'flex',
      flexDirection: 'column',
    }}
  >
    <VirtualGrid {...gridProps} items={null} ItemCard={Card} />
  </Box>
);

export const Loading = LoadingTemplate.bind({});

const EmptyTemplate: ComponentStory<typeof VirtualGrid> = args => (
  <Box
    sx={{
      height: '100vh',
      display: 'flex',
      flexDirection: 'column',
    }}
  >
    <VirtualGrid {...gridProps} items={[]} ItemCard={Card} />
  </Box>
);

export const Empty = EmptyTemplate.bind({});
