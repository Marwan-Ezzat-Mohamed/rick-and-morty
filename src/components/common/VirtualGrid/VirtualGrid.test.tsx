import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import VirtualGrid from './VirtualGrid';
import { mockCharacters } from './../../../__mocks__/characters.mock';

const testData = Array.from({ length: 1000 }, (_, i) => ({
  index: i,
}));
const TestItem = ({ index }: { index: number }) => {
  return <div>{index}</div>;
};

describe('VirtualGrid', () => {
  it('matches snapshot with data', () => {
    const isItemLoaded = jest.fn();
    const loadMoreItems = jest.fn();
    const rendered = render(
      <VirtualGrid
        items={testData}
        ItemCard={TestItem}
        itemHeight={480}
        itemWidth={255}
        gap={25}
        loadMoreItems={loadMoreItems}
        isItemLoaded={isItemLoaded}
        hasMore={false}
      />,
    );

    expect(rendered).toMatchSnapshot();
  });

  it('matches snapshot with no data', () => {
    const isItemLoaded = jest.fn();
    const loadMoreItems = jest.fn();
    const rendered = render(
      <VirtualGrid
        items={[]}
        ItemCard={TestItem}
        itemHeight={495}
        itemWidth={255}
        gap={25}
        loadMoreItems={loadMoreItems}
        isItemLoaded={isItemLoaded}
        hasMore={false}
      />,
    );

    expect(rendered).toMatchSnapshot();
  });

  it('matches snapshot with null data', () => {
    const isItemLoaded = jest.fn();
    const loadMoreItems = jest.fn();
    const rendered = render(
      <VirtualGrid
        items={null}
        ItemCard={TestItem}
        itemHeight={495}
        itemWidth={255}
        gap={25}
        loadMoreItems={loadMoreItems}
        isItemLoaded={isItemLoaded}
        hasMore={false}
      />,
    );

    expect(rendered).toMatchSnapshot();
  });
});
