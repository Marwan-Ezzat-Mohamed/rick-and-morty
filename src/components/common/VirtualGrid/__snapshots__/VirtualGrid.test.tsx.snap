// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualGrid matches snapshot with data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="overflow: visible; height: 0px; width: 0px;"
      >
        <div
          aria-label="grid"
          aria-readonly="true"
          class="ReactVirtualized__Grid ReactVirtualized__List undefined virtual-grid"
          role="grid"
          style="box-sizing: border-box; direction: ltr; height: 0px; position: relative; width: 0px; will-change: transform; overflow-x: hidden; overflow-y: auto;"
          tabindex="0"
        />
      </div>
      <div
        class="resize-triggers"
      >
        <div
          class="expand-trigger"
        >
          <div
            style="width: 1px; height: 1px;"
          />
        </div>
        <div
          class="contract-trigger"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="overflow: visible; height: 0px; width: 0px;"
    >
      <div
        aria-label="grid"
        aria-readonly="true"
        class="ReactVirtualized__Grid ReactVirtualized__List undefined virtual-grid"
        role="grid"
        style="box-sizing: border-box; direction: ltr; height: 0px; position: relative; width: 0px; will-change: transform; overflow-x: hidden; overflow-y: auto;"
        tabindex="0"
      />
    </div>
    <div
      class="resize-triggers"
    >
      <div
        class="expand-trigger"
      >
        <div
          style="width: 1px; height: 1px;"
        />
      </div>
      <div
        class="contract-trigger"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`VirtualGrid matches snapshot with no data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
        No results found
      </h1>
    </div>
  </body>,
  "container": <div>
    <h1>
      No results found
    </h1>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`VirtualGrid matches snapshot with null data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
        Loading...
      </h1>
    </div>
  </body>,
  "container": <div>
    <h1>
      Loading...
    </h1>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
